src/main/java/org/dswarm/wikidataimporter/WikidataDswarmImporter.java

```diff
@@ -84,28 +168,78 @@ private void processGDMResource(final Resource resource) {
 
 			for (final Statement gdmStatement : gdmStatements) {
 
+				statementCount.incrementAndGet();
+
+				final String predicateURI = gdmStatement.getPredicate().getUri();
+
+				if (!wikidataStatementsMap.containsKey(predicateURI)) {
+
+					final List<org.wikidata.wdtk.datamodel.interfaces.Statement> wikidataStatements = new ArrayList<>();
+
+					wikidataStatementsMap.put(predicateURI, wikidataStatements);
+				}
+
 				final Optional<org.wikidata.wdtk.datamodel.interfaces.Statement> optionalWikidataStmt = processGDMStatement(gdmStatement);
 
 				if (!optionalWikidataStmt.isPresent()) {
 
-					// TODO: log non-created statements
+					// log non-created statements
+					LOG.debug("could not process statement '{}'", printGDMStatement(gdmStatement));
 
 					continue;
 				}
 
 				final org.wikidata.wdtk.datamodel.interfaces.Statement wikidataStmt = optionalWikidataStmt.get();
 
-				wikidataStatements.add(wikidataStmt);
+				wikidataStatementsMap.get(predicateURI).add(wikidataStmt);
+
+				processedStatementCount.incrementAndGet();
+
+				final long currentStatementCount = statementCount.get();
+
+				if (currentStatementCount / 10000 == bigStatementCount.get()) {
+
+					bigStatementCount.incrementAndGet();
+
```

Comment:
I'm not sure really how concurrent this code is run or how relevant it is in the end, but there is be a race here. Multiple threads can arrive at the point to check to if condition. If Thread1 is suspended after evaluating the check and before incrementing bigStatementCount, Thread2 can succeed in the same check, increment the counter and then Thread1 increments the counter again. You need a CAS loop here

```
boolean needsUpdate = true;
do {
  final long currentBigStatementCount = bigStatementCount.get();
  needsUpdate = currentStatementCount / 10000 == currentBigStatementCount;
  if (needsUpdate) {
    needsUpdate = !bigStatementCount.compareAndSet(currentBigStatementCount, currentBigStatementCount + 1);
  }
} while (needsUpdate)
```